pennylane/optimize/qnspsa.py

```diff
@@ -0,0 +1,443 @@
+
+        self._update_tensor(tensor_avg)
+        params_next = self._get_next_params(args, grad_avg)
+
+        if len(params_next) == 1:
+            return params_next[0]
+        return params_next
+
+    def _post_process_grad(self, grad_raw_results, grad_dirs):
+        r"""Post process the gradient tape results to get the SPSA gradient estimation.
+
+        Args:
+            grad_raw_results list[np.array]: list of the two qnode results with input parameters
+            perturbed along the ``grad_dirs`` directions
+            grad_dirs list[np.array]: list of perturbation arrays along which the SPSA
+            gradients are estimated
+
+        Returns:
+            list[np.array]: list of gradient arrays. Each gradient array' dimension matches
+            the shape of the corresponding input parameter
+        """
+        loss_plus, loss_minus = grad_raw_results
+        grad = [
+            (loss_plus - loss_minus) / (2 * self.finite_diff_step) * grad_dir
+            for grad_dir in grad_dirs
+        ]
+        return grad
+
+    def _post_process_tensor(self, tensor_raw_results, tensor_dirs):
+        r"""Post process the corresponding tape results to get the metric tensor estimation.
+
+        Args:
+            tensor_raw_results list[np.array]: list of the four perturbed qnode results to compute
+            the estimated metric tensor
+            tensor_dirs list[np.array]: list of the two perturbation directions used to compute
+            the metric tensor estimation. Perturbations on the different input parameters have
+            been concatenated
+
+        Returns:
+            np.array: estimated Fubini-Study metric tensor
+        """
+        tensor_raw_results = [result.squeeze() for result in tensor_raw_results]
+        # For each element of tensor_raw_results, the first dimension is the measured probability in
+        # the computational ket{0} state, which equals the state overlap between the perturbed and
+        # unperturbed circuits.
+        tensor_finite_diff = (
+            tensor_raw_results[0][0]
+            - tensor_raw_results[1][0]
+            - tensor_raw_results[2][0]
+            + tensor_raw_results[3][0]
+        )
+        metric_tensor = (
+            -(
+                np.tensordot(tensor_dirs[0], tensor_dirs[1], axes=0)
+                + np.tensordot(tensor_dirs[1], tensor_dirs[0], axes=0)
+            )
+            * tensor_finite_diff
+            / (8 * self.finite_diff_step**2)
+        )
+        return metric_tensor
+
+    def _get_next_params(self, args, gradient):
+        params = []
+        for arg in args:
+            if not getattr(arg, "requires_grad", False):
+                continue
+            # if an input parameter is a zero-dimension array, add one dimension to form an array.
+            # The returned result will not drop this added dimension
+            if arg.shape == ():
+                arg = arg.reshape(-1)
+            params.append(arg)
```

Comment:
This method returns the updated parameters only if they are trainable. 

Question: does it make more sense to return _all_ of the parameters, but leave the non-trainable ones the same? (I don't know if there's a convention here.)

Reply:
I think what you suggested makes more sense!

Reply:
Right, if there are any edge cases caused by the caveats, then I'd agree that it's fine to address them in another PR.